### PR TITLE
More properly handle nullability of types/literals in Substrait

### PR DESCRIPTION
**Pull request by Blizzara**
## Which issue does this PR close?

Extracted from #10531

Builds on top of #10622 so I'll rebase this once it's merged.

## Rationale for this change

More closely propagate null-ability through Substrait conversions

## What changes are included in this PR?

Handles the nullability of struct fields during Substrait conversions. This isn't perfect; some things are still assumed to just always be nullable (e.g. Literal list elements).

## Are these changes tested?

Tested through round-trip unit tests

## Are there any user-facing changes?

Nulls are handled more accurately

### CODE REVIEW

**Comment by Blizzara** [comment]
@alamb @jonahgao - this next one is up for review/merge now :) And thanks for merging the struct one!

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 1]
```diff
@@ -1686,7 +1691,7 @@ fn to_substrait_bounds(window_frame: &WindowFrame) -> Result<(Bound, Bound)> {
     ))
 }
 
-fn to_substrait_literal(value: &ScalarValue) -> Result<Literal> {
+fn to_substrait_literal(value: &ScalarValue, nullable: bool) -> Result<Literal> {
```
Comment (jonahgao) [inline datafusion/substrait/src/logical_plan/producer.rs:1694]: According to [ExprSchemable::nullable()](https://github.com/apache/datafusion/blob/3e4e09a7b086009921588d52be577ada0ed6d6b9/datafusion/expr/src/expr_schema.rs#L267), the nullability of a literal depends on its `ScalarValue`,  so I think the `nullable` argument for this function could be omitted.
```rust
fn to_substrait_literal(value: &ScalarValue, nullable: bool) -> Result<Literal> {
    if value.is_null() {
        return Ok(Literal {
            nullable: true,
            type_variation_reference: DEFAULT_TYPE_REF,
            literal_type: Some(LiteralType::Null(to_substrait_type(
                &value.data_type(),
                true,
            )?)),
        });
    }
```
At the same time, we can also remove `try_to_substrait_null`.

File: datafusion/substrait/src/logical_plan/consumer.rs [hunk 2]
```diff
@@ -1175,6 +1180,47 @@ pub(crate) fn from_substrait_type(dt: &substrait::proto::Type) -> Result<DataTyp
...
 
+fn is_substrait_type_nullable(dtype: &Type) -> Result<bool> {
+    fn is_nullable(nullability: i32) -> bool {
+        nullability != substrait::proto::r#type::Nullability::Required as i32
+    }
+
+    let nullable = match dtype
+        .clone()
+        .kind
+        .ok_or(substrait_datafusion_err!("Type must contain Kind"))?
```
Comment (jonahgao) [inline datafusion/substrait/src/logical_plan/consumer.rs:1191]:
```suggestion
    let nullable = match dtype
        .kind
        .as_ref()
        .ok_or_else(|| substrait_datafusion_err!("Type must contain Kind"))?
```
Avoid cloning and creating DataFusionError.

File: datafusion/substrait/src/logical_plan/producer.rs (hunk 1 above)
Comment (Blizzara) [reply datafusion/substrait/src/logical_plan/producer.rs]: I think you're right, thanks! That does indeed simplify things. Fixed in [cbd2264](https://github.com/apache/datafusion/pull/10640/commits/cbd22641d3673beec2cb045d658f8c585724b793)

**Review by jonahgao — APPROVED** [review]
Looks good to me 👍

**Comment by jonahgao** [comment]
I plan to merge this PR now as it might conflict with #10646.

**Comment by alamb** [comment]
Thanks @Blizzara and @jonahgao

**Pull request merged**